### PR TITLE
feature[next]: Add argument types to ITIR when tracing

### PR DESCRIPTION
**Pull request by tehrengruber**
Enhances the ITIR fencil tracer by populating the `dtype`, `kind` attributes of the fencil parameters (i.e. `itir.Sym`s) for all arguments that are fields. This is required by the Temporary pass in #1271. Since the ITIR type system is to be refactored only the minimal functionality needed is implemented here.

### CODE REVIEW

File: src/gt4py/next/iterator/tracing.py [hunk 1]
```diff
@@ -248,37 +251,82 @@ def closure(domain, stencil, output, inputs):
...
         return FencilDefinition(
             id=fun.__name__,
             function_definitions=TracerContext.fundefs,
-            params=list(Sym(id=param) for param in param_names),
+            params=params,
             closures=TracerContext.closures,
         )
-
-
-def fendef_tracing(fun, *args, **kwargs) -> FencilDefinition:
```
Comment (tehrengruber) [inline src/gt4py/next/iterator/tracing.py:283]: This function just dropped the keyword arguments. I removed it and made it explicit in the `FendefDispatcher`.

File: tests/next_tests/integration_tests/feature_tests/iterator_tests/test_builtins.py [hunk 2]
```diff
@@ -320,47 +320,39 @@ def test_can_deref(program_processor, stencil):
...
-        (int64(2147483648), int64, int64(2147483648)),  # int64 does not accidentally down-cast
+        (float64("0.1"), float32, np.float32),
+        (int64(42), bool, builtins.bool),
+        (int64(2147483648), int32, np.int32),
+        (int64(2147483648), int64, np.int64),
     ],
 )
 @pytest.mark.parametrize("as_column", [False, True])
-def test_cast(program_processor, as_column, input_value, dtype, expected_value):
+def test_cast(program_processor, as_column, input_value, dtype, np_dtype):
```
Comment (havogt) [inline tests/next_tests/integration_tests/feature_tests/iterator_tests/test_builtins.py:333]: nice

File: tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/fvm_nabla_setup.py [hunk 3]
```diff
@@ -84,10 +84,10 @@ def _is_pole_edge(e, edge_flags):
     def is_pole_edge_field(self):
         edge_flags = np.array(self.mesh.edges.flags())
 
-        pole_edge_field = np.zeros((self.edges_size,))
+        pole_edge_field = np.zeros((self.edges_size,), dtype=bool)
         for e in range(self.edges_size):
             pole_edge_field[e] = self._is_pole_edge(e, edge_flags)
-        return edge_flags
+        return pole_edge_field
```
Comment (havogt) [inline tests/next_tests/integration_tests/multi_feature_tests/iterator_tests/fvm_nabla_setup.py:90]: is this not used?

File: src/gt4py/next/iterator/tracing.py [hunk 4]
```diff
@@ -248,37 +251,82 @@ def closure(domain, stencil, output, inputs):
...
+        if i < len(param_infos):
+            param_info = param_infos[i]
+        else:
+            if param_info.kind != inspect.Parameter.VAR_POSITIONAL:
+                # the last parameter info might also be a keyword or variadic keyword argument, but
+                # they are not supported.
+                raise NotImplementedError(
+                    "Only `POSITIONAL_OR_KEYWORD` or `VAR_POSITIONAL` parameters are supported."
+                )
+            param_info = param_infos[-1]
```
Comment (havogt) [inline src/gt4py/next/iterator/tracing.py:280]: - This is a bit complicated to follow, but I don't have a proposal.
- What's the error if `len(args) < len(positional_params)`?
- Do you know where we actually need `VAR_POSITIONAL`? Probably in the on-the-fly fencil implementation and in some generic tests?

**Review by havogt — COMMENTED** [review]
Just a question...

**Review by havogt — APPROVED** [review]

**Pull request merged**